### DISCUSSION TITLE
Has anyone used xstate for domain modeling?

### DISCUSSION THREAD

**Discussion by zacharyweidenbach** [statelyai/xstate#3398] [General]
In Domain Driven Development (DDD), there are often domain entities that have complex lifecycles which can be difficult to manage or express to non-technical members of the team. These entities often bear a lot of resemblance to state machines. I'm curious if anyone has used xstate and its state charts to articulate the essential complexity of their domain entities as well as to protect against invariants within their systems.

You could imagine a state machine being handed a domain entity, and then synchronizing its state to that of the entity. And then, in order to operate on the entity, the state machine acts as a gatekeeper of sorts, and mutations of the domain entity are only allowed by sending events to the machine. You may want to treat interactions with the machine as an infrastructure concern. The other thing I'm not sure how you'd solve is catching illegal operations sent to the machine. If you send an event to a machine that isn't in the state it needs to be in order to consume the event, nothing happens. But in this example, you may want to detect this and handle it as some kind of error.

Here is a contrived example of how you could express the complex lifecycle of a "work order" from its creation to when payment is received and it is considered "complete".

![Screen Shot 2022-06-10 at 9 06 00 AM](https://user-images.githubusercontent.com/14946412/173106727-2bcf190d-cad8-4957-81b5-986984bf0249.png)

**Comment by davidkpiano** [comment]
DDD is a very interesting topic where I think state machines fit naturally into, and admittedly one that I need to research more!

I agree that state machines are a great way to represent the state of individual entities, and statecharts in particular can model how these entities _interact_ with each other, by sending messages.

To answer this question:

> The other thing I'm not sure how you'd solve is catching illegal operations sent to the machine.

What you can do is add a wildcard transition that will catch unknown events in the root-level state:

![CleanShot 2022-06-12 at 08 48 27@2x](https://user-images.githubusercontent.com/1093738/173234024-b9a376ff-eeff-4509-8109-db2ed60cb767.png)

This is the recommended pattern, and gives you full control over what happens with unrecognized events.

**Reply by williamrjribeiro** [reply]
I'm also interested in this topic. DDD was really effective in the projects I worked.

I recently used the Stately Visual Editor to capture the output of an Event Storming session and it worked really well.

The business process, modelled in the session, flows through multiple bounded contexts. Many state transitions are domain events from different subsystems/services.

My question is: I have a state chart but where should its actor "live"? Services? Aggregates? Entities? Somewhere else? 💭
